File: 0501~1000/0847. Shortest Path Visiting All Nodes/main.py
"""
Source: https://rb.gy/na3fc
Date: 2023/8/13
Skill:
Ref:
Runtime: 2659 ms, faster than 5.01%
Memory Usage: 22.78 MB, less than 27.14%
Time complexity:
Space complexity:
Constraints:
    1 <= graph.length <= 12
"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def shortestPathLength(self, graph: List[List[int]]) -> int:
        n, res = len(graph), sys.maxsize
        target, visited = (1 << n) - 1, defaultdict(int)

        def dfs(pos, cur_state, step):
            nonlocal res
            if cur_state == target:
                res = min(res, step)
                return
            for neighbor in graph[pos]:
                next_state = cur_state
                if next_state & (1 << neighbor) == 0:
                    next_state += (1 << neighbor)
                if (next_state, neighbor) not in visited or step + 1 < visited[(next_state, neighbor)]:
                    visited[(next_state, neighbor)] = step + 1
                    dfs(neighbor, next_state, step + 1)

        for i in range(n):
            start = 1 << i
            visited[(start, i)] = 0
            dfs(i, start, 0)
        return res


if __name__ == "__main__":
    s = Solution()
    graph = [[1, 2, 3], [0], [0], [0]]
    res = s.shortestPathLength(graph)
    print(res)

File: 0501~1000/0847. Shortest Path Visiting All Nodes/main.cpp
/**
 * Source: https://rb.gy/yzwus
 * Date: 2023/9/17
 * Skill:
 * Runtime: 19 ms, faster than 74.80% of C++ online submissions
 * Memory Usage: 9.6 MB, less than 81.65% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *      1 <= graph.length <= 12
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>

using namespace std;

#define ll long long
typedef pair<int, int> pairs;
typedef unsigned long long ULL;
using PULL = pair<ULL, ULL>;


class Solution {
public:
    int shortestPathLength(vector<vector<int>>& graph) {
        int n = graph.size(), step = -1, final_state = 0;
        vector<vector<bool>> visited(n, vector<bool>(1<<n, false));
        deque<pairs> q;
        for (int i = 0; i < n; i++) {
            final_state += 1<<i;
            visited[i][1<<i] = true;
            q.push_back({i, 1<<i});
        }
        vector<unordered_set<int>> adj_set(n);
        for (int i = 0; i < n; ++i) {
            for (auto j: graph[i]) {
                adj_set[i].insert(j);
                adj_set[j].insert(i);
            }
        }
        while (!q.empty()) {
            int sz = q.size();
            step++;
            for (int i = 0; i < sz; ++i) {
                auto [pos, state] = q.front(); q.pop_front();
                if (state == final_state)
                    return step;
                for (auto j: adj_set[pos]) {
                    int next_state = state | (1<<j);
                    if (!visited[j][next_state]) {
                        visited[j][next_state] = true;
                        q.push_back({j, next_state});
                    }
                }
            }
        }
        return -1;
    }
};

int main() {
    Solution s;
    vector<int> nums{8,7,3,5,7,2,4,9};
    vector<vector<int>> arr{{0,1,0,0,0}, {0,1,0,1,0}, {0,0,0,1,0}};
}